router/core/src/main/scala/io/buoyant/router/RoutingFactory.scala

```diff
@@ -3,22 +3,26 @@ package io.buoyant.router
 import com.twitter.finagle.{param => _, _}
 import com.twitter.finagle.buoyant.{Dst, DstBindingFactory}
 import com.twitter.finagle.tracing.Trace
-import com.twitter.util.{Future, Time}
+import com.twitter.util.{Throw, Return, Future, Time}
 import scala.util.control.NoStackTrace
 
 object RoutingFactory {
   val role = Stack.Role("RoutingFactory")
   val description = "Performs per-request name binding"
 
-  case class UnknownDst[Req](request: Req, cause: Throwable)
-    extends Exception(s"Unknown destination: $request / ${cause.getMessage}", cause)
+  case class UnknownDst[Req](request: Req, reason: String)
+    extends Exception(s"Unknown destination: $request / $reason")
     with NoStackTrace
 
+  sealed trait RequestIdentification[Req]
+  class IdentifiedRequest[Req](val dst: Dst, val request: Req) extends RequestIdentification[Req]
+  class UnidentifiedRequest[Req](val reason: String, val request: Req) extends RequestIdentification[Req]
```

Comment:
tioli: you can provide unapply helpers for these (and we'll just never update them... or have to break API again). This is, for instance, how the Name ADT works: the Name types are classes that have companion objects with unapply